awx/main/management/commands/host_metric.py

```diff
@@ -1,26 +1,167 @@
 from django.core.management.base import BaseCommand
 import datetime
 from django.core.serializers.json import DjangoJSONEncoder
-from awx.main.models.inventory import HostMetric
+from awx.main.models.inventory import HostMetric, HostMetricSummaryMonthly
+from awx.main.analytics.collectors import config
 import json
+import sys
+import tempfile
+import tarfile
+import csv
+
+PREFERRED_ROW_COUNT = 500000
 
 
 class Command(BaseCommand):
     help = 'This is for offline licensing usage'
 
+    def host_metric_queryset(self, result, offset=0, limit=PREFERRED_ROW_COUNT):
+        list_of_queryset = list(
+            result.values(
+                'id',
+                'hostname',
+                'first_automation',
+                'last_automation',
+                'last_deleted',
+                'automated_counter',
+                'deleted_counter',
+                'deleted',
+                'used_in_inventories',
+            ).order_by('first_automation')[offset : offset + limit]
+        )
+
+        return list_of_queryset
+
+    def host_metric_summary_monthly_queryset(self, result, offset=0, limit=PREFERRED_ROW_COUNT):
+        list_of_queryset = list(
+            result.values(
+                'id',
+                'date',
+                'license_consumed',
+                'license_capacity',
+                'hosts_added',
+                'hosts_deleted',
+                'indirectly_managed_hosts',
+            ).order_by(
+                'date'
+            )[offset : offset + limit]
+        )
+
+        return list_of_queryset
+
+    def paginated_db_retrieval(self, type, filter_kwargs, offset=0, limit=PREFERRED_ROW_COUNT):
+        list_of_queryset = []
+        if type == 'host_metric':
+            result = HostMetric.objects.filter(**filter_kwargs)
```

Comment:
there could be also fine-grained db pagination in case of big rows_per_file settings (i.e. 10000)

This can produce `SELECT * from main_hostmetric WHERE ... LIMIT 500000`. That's quite a big number for db.

Reply:
So `--rows_per_file`  is the parameter for regulating how many rows to fetch at any given time.



- `For awx-manage host_metric --csv --rows_per_file 10000`, 
we will keep fetching 10000 rows until we have fetched all rows and then display the csv in standard output

- `For awx-manage host_metric --tarball --rows_per_file 10000`, 
we will fetch 10000 rows and create the tarball with several individual CSV files where each CSV would have up to 10000 rows

Reply:
yes, it's a parameter how big should be the CSV file. 
But my concern is that if we default 500k, then *default* query to the database is `SELECT .. LIMIT 500k`, that sounds dangerous. So I'm suggesting to make independent db pagination and csv pagination

Reply:
Martin is correct, unless you'll do `COPY` directly to csv (that should be batched operation), this might blow up in memory. E.g. we do this in Analytics Exporter, COPY to csv without any explicit batching (postgres takes care of that), then batch reading of the csv to convert it to different format. (so we keep the memory low) That should be the most effective way to do it.


But paginated read from DB should work ok-ish for now. (something like hardcoded page of 10k rows we load from DB at once and append to the csv)

Reply:
> So I'm suggesting to make independent db pagination and csv pagination

I had thought about this, but was trying to avoid this situation because two pagination schemes are just going to add some complexity.

I would have to evaluate this to make a decision as we do not want to overcomplicate the code.

------------------

The easiest approach here  would be to set default value of `--rows_per_file` to 10000

Reply:
Added functionality for fetching rows in batches of 10k and creating CSVs with whatever is specified in `--rows_per_file`
Since 500k is a really large number (even for the postgres function), its best to use 10k-chunking approach